File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>>ans; //2d vector to store ans
    void solve(int i,vector<int>& arr,vector<int>&temp,int target) {
        if(target==0) {
            ans.push_back(temp);
            return;
        }
        if(target<0) {
            return;
        }
        if(i==arr.size()){
            return;
        }
        solve(i+1,arr,temp,target);
        temp.push_back(arr[i]);
        solve(i,arr,temp,target-arr[i]);
        temp.pop_back(); //backtrack
    }
        
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        ans.clear();
        vector<int> temp;
        solve(0,candidates,temp,target);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def dfs(i,current,value):
            if value==target:
                res.append(current.copy())
                return
            if i>=len(candidates) or value>target:
                return
            current.append(candidates[i])
            dfs(i,current,value+candidates[i])
            current.pop()
            dfs(i+1,current,value)
            
        dfs(0,[],0)
        return res